tensorflow/compiler/tf2xla/kernels/image_resize_ops.cc

```diff
@@ -555,13 +561,22 @@ void GeneralCompile(XlaOpKernelContext* ctx, bool align_corners_,
     }
   }
 
+  // Bilinear always outputs float, but nearest neighbor keeps the original type
+  if (!is_kernel_bilinear && original_input_type != input_type) {
+    input = xla::ConvertElementType(input, original_input_type);
+  }
   ctx->SetOutput(0, input);
 }
 
+bool static IsGPUDevice(OpKernelConstruction* ctx) {
+  absl::string_view s = ctx->device_type().type_string();
+  return s == DEVICE_GPU_XLA_JIT || s == DEVICE_XLA_GPU;
+}
+
 class ResizeNearestNeighborOp : public XlaOpKernel {
  public:
   explicit ResizeNearestNeighborOp(OpKernelConstruction* ctx)
-      : XlaOpKernel(ctx) {
+      : XlaOpKernel(ctx), is_gpu_device_(IsGPUDevice(ctx)) {
```

Comment:
Save device info on OP. I found an existing example from stateless_random_ops.cc. An alternative way is to register OPs with different/specific constraints (type and device), similar to those in argmax_op.cc. I have tried both approaches and both work. I am open to either option

Reply:
Since the first approach is implemented already, let's go with it, but I will defer the final judgment to @blakehechtman (hmm @tjablin doesn't show up when I type).

Reply:
Seems okay to me.